### PR TITLE
Adding new Ceph for Windows guide in XML format

### PR DESCRIPTION
**Pull request by asettle**
This was written by Mike Latimer and is released for a technical preview as a standalone guide.

### CODE REVIEW

File: xml/windows-ceph.xml [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+  <dm:docmanager xmlns:dm="urn:x-suse:ns:docmanager">
+   <dm:editurl>https://github.com/SUSE/doc-ses/edit/master/xml/</dm:editurl>
+   <dm:translation>yes</dm:translation>
+   <dm:release>SES 7</dm:release>
+  </dm:docmanager>
+ </info>
+ <section xml:id="intro-windows">
+   <title>Introduction</title>
+   <para>
+     Ceph is a highly resilient software defined storage offering, which has
```
Comment (lproven) [inline xml/windows-ceph.xml:19]:
```suggestion
     Ceph is a highly-resilient software-defined-storage offering, which has
```

File: xml/windows-ceph.xml [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+     only been available to Windows environments through the use of iSCSI or
+     CIFS gateways. This gateway architecture introduces single points of
+     contacts and limits both fault tolerance and bandwidth - compared to the
```
Comment (lproven) [inline xml/windows-ceph.xml:22]:
```suggestion
     contacts, and limits both fault-tolerance and bandwidth compared to the
```

Comment (lproven) [reply xml/windows-ceph.xml:22]: "points of contacts" plural?

File: xml/windows-ceph.xml [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+     native I/O paths of Ceph with RADOS.
+   </para>
+   <para>
+     In order to bring the benefits of native Ceph to Windows environments,
+     USE partnered with Cloudbase Solutions to port Ceph to the Windows
```
Comment (lproven) [inline xml/windows-ceph.xml:27]:
```suggestion
     SUSE partnered with Cloudbase Solutions to port Ceph to the Windows
```

File: xml/windows-ceph.xml [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+     (The Tech Preview version is not yet signed, but will be available at
+     a later date.)
+   </para>
+   <para>
+     Once installed, the 'WNBD SCSI Virtual Adapter' driver can be seen in
+     Device Manager as a Storage controller. Multiple adapters may be seen
+     to handle multiple number of RBD connections.
+   </para>
+   <para>
+     The 'rbd' command is used to create, remove, import, export, map or unmap
```
Comment (lproven) [inline xml/windows-ceph.xml:137]:
```suggestion
     The <command>rbd</command> command is used to create, remove, import, export, map or unmap
```

File: xml/windows-ceph.xml [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+     images exactly like it is used on Linux.
+   </para>
+   <section xml:id="mapping-images">
+     <title>Mapping Images</title>
+     <para>
+       The behavior of the 'rbd' command is similar to the Linux counterpart,
```
Comment (lproven) [inline xml/windows-ceph.xml:143]:
```suggestion
       The behavior of the <command>rbd</command> command is similar to its Linux counterpart,
```

File: xml/windows-ceph.xml [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+         <para>
+           Device paths cannot be requested. The disk number and path will
+           be picked by Windows. If a device path is provided by the user when
+           mapping an image, it will be used as an identifier, which can also
+           be used when unmapping the image.
+         </para>
+       </listitem>
+       <listitem>
+         <para>
+           The 'show' command was added, which describes a specific mapping.
```
Comment (lproven) [inline xml/windows-ceph.xml:157]:
```suggestion
           The <command>show</command> command was added, which describes a specific mapping.
```

File: xml/windows-ceph.xml [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+           This can be used for retrieving the disk path.
+         </para>
+       </listitem>
+       <listitem>
+         <para>
+           The 'service' command was added, allowing rbd-nbd to run as a Windows
```
Comment (lproven) [inline xml/windows-ceph.xml:163]:
```suggestion
           The <command>service</command> command was added, allowing <literal>rbd-nbd</literal> to run as a Windows
```

File: xml/windows-ceph.xml [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+           service. All mappings are currently persistent, being recreated when
+           the service stops, unless explicitly unmapped. The service disconnects
+           the mappings when being stopped.
+         </para>
+       </listitem>
+       <listitem>
+         <para>
+           The 'list' command also includes a 'status' column.
```
Comment (lproven) [inline xml/windows-ceph.xml:171]:
```suggestion
           The <command>list</command> command also includes a <literal>status</literal> column.
```

File: xml/windows-ceph.xml [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+     In order to ensure that rbd-nbd mappings survive host reboots, a new
+     Windows service, called "Ceph RBD Mapping Service", has been created.
+     This service automatically maintains mappings as they are added using
+     the Ceph tools. All mappings are currently persistent, being recreated
+     when the service starts, unless explicitly unmapped. The service
+     disconnects all mappings when being stopped.
+   </para>
+   <para>
+     This service also adjust the Windows service start order so that rbd
+     images can be mapped before starting services that may depend on it, such as VMS.
```
Comment (lproven) [inline xml/windows-ceph.xml:256]:
```suggestion
     images can be mapped before starting services that may depend on it, such as VMs.
```

File: xml/windows-ceph.xml [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+ </section>
+ <section xml:id="windows-cephfs">
+   <title>CephFS</title>
+   <para>
+     Ceph for Windows provides CephFS support through the Dokany FUSE wrapper.
+     In order to use CephFS, please install Dokany v1.3.1 or newer, which can
+     be found at the following URL: https://github.com/dokan-dev/dokany/releases
+   </para>
+   <para>
+     With Dokany installed, and the ceph.conf and ceph.client.admin.keyring
```
Comment (lproven) [inline xml/windows-ceph.xml:273]:
```suggestion
     With Dokany installed, and the <filename>ceph.conf</filename> and <filename>ceph.client.admin.keyring</filename>
```

File: xml/windows-ceph.xml [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+     configuration files in place, CephFS can be mounted using the
+     ceph-dokan.exe commands, as in the following example:
```
Comment (lproven) [inline xml/windows-ceph.xml:275]:
```suggestion
     <command>ceph-dokan.exe</command> command, as in the following example:
```

File: xml/windows-ceph.xml [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+   </para>
+<screen>
+ceph-dokan.exe -l x
+</screen>
+   <para>
+     The above command will mount the default ceph filesystem using the drive
+     letter 'x'. If 'ceph.conf' is not placed at the default location
```
Comment (lproven) [inline xml/windows-ceph.xml:282]:
```suggestion
     letter 'x'. If <filename>ceph.conf</filename> is not placed at the default location
```

File: xml/windows-ceph.xml [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+     (which is 'C:\ProgramData\ceph\ceph.conf', a '-c' parameter can be used
```
Comment (lproven) [inline xml/windows-ceph.xml:283]:
```suggestion
     (which is <filename>C:\ProgramData\ceph\ceph.conf</filename>), a <option>-c</option> parameter can be used
```

File: xml/windows-ceph.xml [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+     to specify the location of ceph.conf.
```
Comment (lproven) [inline xml/windows-ceph.xml:284]:
```suggestion
     to specify the location of <filename>ceph.conf</filename>.
```

File: xml/windows-ceph.xml [hunk 15]
```diff
@@ -0,0 +1,307 @@
...
+   </para>
+   <para>
+     The '-l' argument also allows using an empty folder as a mountpoint instead
+     of a drive letter.
+   </para>
+   <para>
+     The uid and gid used for mounting the filesystem defaults to 0 and may be
```
Comment (lproven) [inline xml/windows-ceph.xml:291]:
```suggestion
     The UID and GID used for mounting the filesystem defaults to <literal>0</literal> and may be
```

File: xml/windows-ceph.xml [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+     changed using the '-u' and '-g' arguments. '-n' can be used in order to
```
Comment (lproven) [inline xml/windows-ceph.xml:292]:
```suggestion
     changed using the <option>-u</option> and <option>-g</option> arguments. <option>-n</option> can be used in order to
```

File: xml/windows-ceph.xml [hunk 17]
```diff
@@ -0,0 +1,307 @@
...
+     skip enforcing permissions on client side. Be aware that Windows ACLs are
+     ignored. Posix ACLs are supported but cannot be modified using the current
```
Comment (lproven) [inline xml/windows-ceph.xml:294]:
```suggestion
     ignored. POSIX ACLs are supported but cannot be modified using the current
```

File: xml/windows-ceph.xml [hunk 18]
```diff
@@ -0,0 +1,307 @@
...
+     CLI. In the future, we may add some command actions to change file
+     ownership or permissions.
+   </para>
+   <para>
+     For debugging purposes, '-d' and 's' may be used. The first one enables
```
Comment (lproven) [inline xml/windows-ceph.xml:299]:
```suggestion
     For debugging purposes, <option>-d</option> and  <option>-s</option> may be used. The former enables
```

File: xml/windows-ceph.xml [hunk 19]
```diff
@@ -0,0 +1,307 @@
...
+     debug output and the latter enables stderr logging. By default, debug
```
Comment (lproven) [inline xml/windows-ceph.xml:300]:
```suggestion
     debug output and the latter enables <literal>stderr</literal> logging. By default, debug
```

File: xml/windows-ceph.xml [hunk 20]
```diff
@@ -0,0 +1,307 @@
...
+     messages are sent to a connected debugger.
+   </para>
+   <para>
+     You may use '--help' to get the full list of available options.
```
Comment (lproven) [inline xml/windows-ceph.xml:304]:
```suggestion
     You may use <option>--help</option> to get the full list of available options.
```

File: xml/windows-appendix.xml [hunk 21]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE appendix
+[
+  <!ENTITY % entities SYSTEM "entity-decl.ent">
+    %entities;
+]>
+<appendix xmlns="http://docbook.org/ns/docbook"
+ xmlns:xi="http://www.w3.org/2001/XInclude"
+ xmlns:xlink="http://www.w3.org/1999/xlink" version="5.0" xml:id="windows-appendix">
+ <title>Appendicies</title>
```
Comment (lproven) [inline xml/windows-appendix.xml:10]:
```suggestion
 <title>Appendices</title>
```

File: xml/windows-appendix.xml [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+  <dm:docmanager xmlns:dm="urn:x-suse:ns:docmanager">
+   <dm:editurl>https://github.com/SUSE/doc-ses/edit/master/xml/</dm:editurl>
+   <dm:translation>yes</dm:translation>
+   <dm:release>SES 7</dm:release>
+  </dm:docmanager>
+ </info>
+ <sect1 xml:id="windows-appendix-a">
+   <title>Appendix A: Sample Configuration Files</title>
+   <para>
+     C:\ProgramData\ceph\ceph.conf
```
Comment (lproven) [inline xml/windows-appendix.xml:21]:
```suggestion
     <filename>C:\ProgramData\ceph\ceph.conf</filename>
```

File: xml/windows-appendix.xml [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+   </para>
+<screen>
+[global] log to stderr = true run dir = C:/ProgramData/ceph/out crash dir = C:/ProgramData/ceph/out [client]
+keyring = C:/ProgramData/ceph/keyring log file = C:/ProgramData/ceph/out/$name.$pid.log admin
+socket = C:/ProgramData/ceph/out/$name.$pid.asok [global] ;
+pecify IP addresses for monitor nodes as in the following example: ;
+mon host = [v2:10.1.1.1:3300,v1:10.1.1.1:6789] [v2:10.1.1.2:3300,v1:10.1.1.2:6789] [v2:10.1.1.3:3300,v1:1.1.1.3:6789]
+</screen>
+   <para>
+     C:\ProgramData\ceph\ceph.client.admin.keyring
```
Comment (lproven) [inline xml/windows-appendix.xml:31]:
```suggestion
     <filename>C:\ProgramData\ceph\ceph.client.admin.keyring</filename>
```

File: xml/windows-appendix.xml [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+<screen>
+; This file should be copied directly from /etc/ceph/ceph.client.admin.keyring
+; The contents should be similar to the following example: [client.admin] key
+= ADCyl77eBBAAABDDjX72tAljOwv04m121v/7yA== caps mds = "allow *" caps mon = "allow *" caps osd = "allow *" caps mgr = "allow *"
+</screen>
+ </sect1>
+ <sect1 xml:id="windows-appendix-b">
+   <title>Appendix B: Upstream Projects</title>
+   <para>
+     The Ceph for Windows effort is being done entire in Open Source, and
```
Comment (lproven) [inline xml/windows-appendix.xml:42]:
```suggestion
     The Ceph for Windows effort is being done entirely in Open Source, and
```

File: xml/windows-appendix.xml [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+     in conjunction with the upstream project(s). For more development level
```
Comment (lproven) [inline xml/windows-appendix.xml:43]:
```suggestion
     in conjunction with the upstream project(s). For more development-level
```

**Comment by asettle** [comment]
@lproven ready for you now :)

File: xml/windows-appendix.xml (hunk 21 above)
Comment (sknorr) [reply xml/windows-appendix.xml]: Acc. to style guide, actually "appendixes", https://documentation.suse.com/style/current/single-html/docu_styleguide/#sec-terminology (though I like the version with c better too)

Comment (asettle) [reply xml/windows-appendix.xml]: I'll roll with that.

File: xml/windows-ceph.xml [hunk 26]
```diff
@@ -0,0 +1,317 @@
...
+ </section>
+ <section xml:id="windows-cephfs">
+   <title>&cephfs;</title>
+   <para>
+     &ceph; for &mswin; provides &cephfs; support through the Dokany FUSE wrapper.
+     In order to use &cephfs;, install Dokany v1.3.1 or newer using the
+     installers available here: https://github.com/dokan-dev/dokany/releases
+   </para>
+   <para>
+     With Dokany installed, <filename>ceph.conf</filename> and <filename>ceph.client.admin.keyring</filename>
```
Comment (lproven) [inline xml/windows-ceph.xml:281]:
```suggestion
     With Dokany installed, and <filename>ceph.conf</filename> and <filename>ceph.client.admin.keyring</filename>
```

File: xml/windows-ceph.xml [hunk 27]
```diff
@@ -0,0 +1,317 @@
...
+     configuration files in place, &cephfs; can be mounted using the
+     <command>ceph-dokan.exe</command> command. For example:
```
Comment (lproven) [inline xml/windows-ceph.xml:283]: I think the ".exe" is implicit and thus superfluous.

File: xml/windows-ceph.xml [hunk 28]
```diff
@@ -0,0 +1,317 @@
...
+   </para>
+<screen>
+ceph-dokan.exe -l x
```
Comment (lproven) [inline xml/windows-ceph.xml:286]: Also here.

File: xml/windows-ceph.xml [hunk 29]
```diff
@@ -0,0 +1,317 @@
...
+</screen>
+   <para>
+     This command mounts the default &ceph; filesystem using the drive
+     letter <literal>x</literal>. If <filename>ceph.conf</filename> is not placed
```
Comment (lproven) [inline xml/windows-ceph.xml:290]:
```suggestion
     letter <literal>X</literal>. If <filename>ceph.conf</filename> is not placed
```

Comment (lproven) [reply xml/windows-ceph.xml:290]: Windows drive letters are not case-sensitive but the UI displays them as capitals.

File: xml/windows-ceph.xml [hunk 30]
```diff
@@ -0,0 +1,317 @@
...
+     </listitem>
+     <listitem>
+       <para>
+         &cephfs;
+       </para>
+     </listitem>
+   </itemizedlist>
+   <para>
+     Additional information on the background on this effort can be found
+     through the following SUSECON Digital session:
```
Comment (tbazant) [inline xml/windows-ceph.xml:45]:
```suggestion
     You can find additional information on the background of this effort
     through the following SUSECON Digital session:
```

File: xml/windows-ceph.xml [hunk 31]
```diff
@@ -0,0 +1,317 @@
...
+       </para>
+     </listitem>
+   </itemizedlist>
+   <para>
+     After installing &ceph; for &mswin;, manual modifications are required to
+     provide access to a &ceph; cluster. The files which must be created or
+     modified are as follows:
+   </para>
+<screen>
+C:\ProgramData\ceph\ceph.conf C:\ProgramData\ceph\ceph.client.admin.keyring
```
Comment (tbazant) [inline xml/windows-ceph.xml:123]:
```suggestion
C:\ProgramData\ceph\ceph.conf
C:\ProgramData\ceph\ceph.client.admin.keyring
```

File: xml/windows-ceph.xml [hunk 32]
```diff
@@ -0,0 +1,317 @@
...
+   <para>
+     With Dokany installed, <filename>ceph.conf</filename> and <filename>ceph.client.admin.keyring</filename>
+     configuration files in place, &cephfs; can be mounted using the
+     <command>ceph-dokan.exe</command> command. For example:
+   </para>
+<screen>
+ceph-dokan.exe -l x
+</screen>
+   <para>
+     This command mounts the default &ceph; filesystem using the drive
```
Comment (tbazant) [inline xml/windows-ceph.xml:289]:
```suggestion
     This command mounts the default &ceph; file system using the drive
```

File: xml/windows-ceph.xml [hunk 33]
```diff
@@ -0,0 +1,317 @@
...
+     at the default location
+     (<filename>C:\ProgramData\ceph\ceph.conf</filename>), a <option>-c</option>
+     parameter can be used to specify the location of <filename>ceph.conf</filename>.
+   </para>
+   <para>
+     The <option>-l</option> argument also allows using an empty folder as a
+     mountpoint instead of a drive letter.
+   </para>
+   <para>
+     The UID and GID used for mounting the filesystem defaults to <literal>0</literal>
```
Comment (tbazant) [inline xml/windows-ceph.xml:300]:
```suggestion
     The UID and GID used for mounting the file system defaults to <literal>0</literal>
```

File: xml/windows-ceph.xml [hunk 34]
```diff
@@ -0,0 +1,317 @@
...
+   </itemizedlist>
+   <para>
+     Additional information on the background on this effort can be found
+     through the following SUSECON Digital session:
+   </para>
+   <itemizedlist>
+     <listitem>
+       <para>
+         &ceph; in a Windows World (TUT-1121) Presenters: Mike Latimer (SUSE)
+         Alessandro Pilotti (Cloudbase Solutions)
```
[the remark this answers is not in the document]
Comment (asettle) [reply xml/windows-ceph.xml:51]: Ongoing discussion - it's a SUSECon link at the moment, so I'm trying to find out if we can link to that, or the youtube. So far the searchable option is okay for the draft :)

File: xml/windows-ceph.xml [hunk 35]
```diff
@@ -0,0 +1,317 @@
...
+     cluster. Sample configuration files are provided in
+     <xref linkend="windows-appendix-a"/>.
+   </para>
+ </section>
+ <section xml:id="rados-block-device">
+   <title>&rbd; (RBD)</title>
+   <para>
+     Support for RBD devices is possible through a combination of &ceph;
+     tools and &mswin; WNBD. This driver is in the process of being Windows Hardware Quality Labs  (WHQL)
+     certified and signed by &suse;.
```
[the remark this answers is not in the document]
Comment (asettle) [reply xml/windows-ceph.xml:136]: Hmm good point. I think the validation is in two separate steps:
1. Driver needs to be validated by WHQL
2. Then SUSE will sign off on the usage of the driver. But I'll verify with Mike.

File: xml/windows-ceph.xml [hunk 36]
```diff
@@ -0,0 +1,317 @@
...
+         <para>
+           The <command>show</command> command was added, which describes a specific mapping.
+           This can be used for retrieving the disk path.
+         </para>
+       </listitem>
+       <listitem>
+         <para>
+           The <command>service</command> command was added, allowing <literal>rbd-nbd</literal> to run as a &mswin;
+           service. All mappings are currently persistent and being recreated when
+           the service stops, unless they are explicitly unmapped. The service disconnects
```
[the remark this answers is not in the document]
Comment (asettle) [reply xml/windows-ceph.xml:172]: `will be` is correct.

File: xml/windows-ceph.xml [hunk 37]
```diff
@@ -0,0 +1,317 @@
...
+<screen>
+SYSTEM\CurrentControlSet\Services\rbd-nbd
+</screen>
+ </section>
+ <section xml:id="windows-cephfs">
+   <title>&cephfs;</title>
+   <para>
+     &ceph; for &mswin; provides &cephfs; support through the Dokany FUSE wrapper.
+     In order to use &cephfs;, install Dokany v1.3.1 or newer using the
+     installers available here: https://github.com/dokan-dev/dokany/releases
```
[the remark this answers is not in the document]
Comment (asettle) [reply xml/windows-ceph.xml:278]: Plain text at this time. I'll need to figure out what's the protocol for github linking in SES docs. I'll come back to it.

File: xml/windows-ceph.xml (hunk 35 above)
[the remark this answers is not in the document]
Comment (asettle) [reply xml/windows-ceph.xml:137]: Alright, talked with Mike. Windows drivers must be "signed" in order to be accepted. (By default, Windows complains when you try to install a signed driver.) The WHQL certification process is a Microsoft test framework you have to pass in order to also get Microsoft to sign and certify the driver. He said that we are doing both, but it's probably fine to just drop the comment about signing by SUSE, as most people will likely already expect that

**Pull request merged**